### PR TITLE
Add user flags to events and places

### LINKED ISSUE #25: Implement user destination flag options
Modify database to store user destination flags: been, want to go, not interested, liked. These options are mutually exclusive.

Update database when user toggles a flag, and display appropriate icons/labels for selected flags on destinations.

Store in a separate table from the destinations, and reference them by identifier, as the destinations may be cleared and reloaded at any time.

### PR DESCRIPTION
**Pull request by maurizi**
## Overview

Store user flags (been, like, etc.) when a user selects a flag for an event/destination from the list/detail screens, and display the correct image for already flagged events & destinations.

## Demo
screenshot_1523470790
screenshot_1523470775
screenshot_1523470763

## Notes

 - User flags are stored in the separate `AttractionFlag` entity, to allow clearing/refetching the Event and Destination tables. Whenever events or destinations are retrieved we also retrieve the appropriate flags and set them on the model(s) for easier access.

Closes #25

### CODE REVIEW

File: app/src/main/java/com/gophillygo/app/data/models/AttractionFlag.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import android.util.SparseArray;
+
+import com.google.gson.annotations.SerializedName;
+import com.gophillygo.app.R;
+
+@Entity
+public class AttractionFlag {
+
+    public enum Option {
+        NotSelected (0, R.drawable.ic_add_black_24dp),
```
Comment (flibbertigibbet) [inline app/src/main/java/com/gophillygo/app/data/models/AttractionFlag.java]: This mapping is for the black icon only. There are also white versions of these icons, and we may add versions on other colors (see wireframes, like [here](https://marvelapp.com/4c295bb/screen/37237000)). At least if we change the color of the current selection in the popover, we'll need the white icon resource.

File: app/src/main/java/com/gophillygo/app/adapters/AttractionListAdapter.java [hunk 2]
```diff
@@ -74,7 +82,10 @@ public void optionsButtonClick(View view, T attraction) {
         PopupMenu menu = new PopupMenu(context, view);
         menu.getMenuInflater().inflate(R.menu.place_options_menu, menu.getMenu());
         menu.setOnMenuItemClickListener(item -> {
-            Log.d(LOG_LABEL, "Clicked " + item.toString());
+            AttractionFlag flag = attraction.createAttractionFlag(item.getItemId());
+            attraction.setFlag(flag);
+            viewModel.updateAttractionFlag(flag);
```
Comment (flibbertigibbet) [inline app/src/main/java/com/gophillygo/app/adapters/AttractionListAdapter.java]: This seems like an action that should be able to be accomplished with one call to set the flag, or maybe two if the update to the `LiveData` source object isn't used.

File: app/src/main/java/com/gophillygo/app/data/DestinationRepository.java [hunk 3]
```diff
@@ -92,21 +121,71 @@ protected Void doInBackground(Void... voids) {
...
 
     public LiveData<Resource<List<Event>>> loadEvents() {
-        return new AttractionNetworkBoundResource<Event>(webservice, destinationDao, eventDao) {
+        LiveData<Resource<List<Event>>> data = new AttractionNetworkBoundResource<Event>(webservice, destinationDao, eventDao) {
             @NonNull @Override
             protected LiveData<List<Event>> loadFromDb() {
                 return eventDao.getAll();
             }
         }.getAsLiveData();
+        return addAttractionFlags(data, true);
```
Comment (flibbertigibbet) [inline app/src/main/java/com/gophillygo/app/data/DestinationRepository.java]: If possible, would prefer to not wrap here with `MutableLiveData`, and rather modify the database queries to include the flag as an attraction property, as is done currently with the related destination name being set on events when queried.

File: app/src/main/java/com/gophillygo/app/data/DestinationRepository.java [hunk 4]
```diff
@@ -92,21 +121,71 @@ protected Void doInBackground(Void... voids) {
...
+    }
+
+    private <T extends Attraction> LiveData<T> addAttractionFlag(LiveData<T> data) {
+        return switchMap(data, attraction -> {
+            MutableLiveData<T> result = new MutableLiveData<>();
+            if (attraction == null) {
+                result.postValue(attraction);
```
Comment (flibbertigibbet) [inline app/src/main/java/com/gophillygo/app/data/DestinationRepository.java]: Should post `null` here to be more explicit (the linter complains about this line).

**Comment by maurizi** [comment]
@flibbertigibbet This is ready for another look. The biggest change is in https://github.com/azavea/cac-tripplanner-android/pull/36/commits/3fb4ca801cef5300b3acf3585e926c0832fadddd which removes the use of `MutableLiveData` and refactors the `Event` and `Destination` classes somewhat.

>  The ViewModel should not be passed to the list adapter constructors, as this is a possible memory leak; references to Android Components are obtained through dependency injection and so should not be passed around. The existing AttractionListItemClickListener could be modified instead (and renamed) to add a second callback to listen to the popover on the activity, which already has a reference to the ViewModel and will better maintain a separation of concerns.

This was fixed in https://github.com/azavea/cac-tripplanner-android/pull/36/commits/da4502ddf571d872b9900c689428441b6c30f036

> After interacting with a list card's flags, the list scroll position resets to the head of the list. This may be due to changes to the data causing the source to change.

This was fixed in https://github.com/azavea/cac-tripplanner-android/pull/36/commits/8a2c972edf04b4adbb01d9e4c42fe2728ed5e593

> Popover current item selection should be highlighted, as in the wireframe.
I opened issue #38 to tackle this

I opened #38 to address this.

**Pull request merged**